runtime/lua/vim/lsp.lua

```diff
@@ -811,78 +811,113 @@ local function once(fn)
   end
 end
 
+--@private
+--- converts retrieves buffer text based on row and column range
+---
+--@param bufnr integer Buffer handle
+--@param range table table containing start_row, start_column, new_end_row, and new_end_column from on_bytes callback
+--@returns string text from buffer
+local function buf_range_to_text(bufnr, range)
+  local lines = vim.api.nvim_buf_get_lines(bufnr, range.start_row, range.start_row + range.new_end_row + 1, false)
+  for idx = 1, #lines do
```

Comment:
I don't think `\n` on final newline should be there. We are extracting one extra line to get the `new_end_column` part of the final line. so we can use `table.concat(lines, '\n')` in the end instead.

Reply:
I had this originally, but it wasn't representing the test correctly for single empty line additions (like `o`)

Reply:
Hmm are you sure? with `o` new_end_row should be 1 so we get one more line, is this not the case?

Reply:
Here's what happens when I use table.concat

https://user-images.githubusercontent.com/13316262/108682719-84b7dc80-74a5-11eb-84f5-c8263bd297ba.mov

Reply:
This is what happens with the PR as it currently is:

https://user-images.githubusercontent.com/13316262/108683109-060f6f00-74a6-11eb-8366-114ddd70625b.mov

Reply:
That's weird because the updates for the latter looks wrong: it inserts two extra lines upon undo.

Reply:
It's definitely still wrong, I'm just trying to figure out exactly how it is wrong 😆 Adding and undoing 3 lines breaks it